File: 0424_characterReplacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        hashmap = {}
        left = 0
        maximum = 0
        output = 0
        for right, char in enumerate(s):
            hashmap[char] = 1 + hashmap.get(char, 0)
            maximum = max(maximum, hashmap[char])
            if right - left - maximum >= k:
                hashmap[s[left]] -= 1
                left += 1
            output = max(output, right - left + 1)
        return output

File: 0001-0500/0424_characterReplacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        int[] charMap = new int[26];
        int longest = 0;
        int left = 0;
        int right = 0;
        int max;
        int sum;

        while (right < s.length()) {
            max = getMax(charMap);
            sum = getSum(charMap);
            longest = Math.max(longest, sum);

            if (sum - max < k) {
                charMap[s.charAt(right) - 'A']++;
                right++;
            } else if (charMap[s.charAt(right) - 'A'] == max) {
                charMap[s.charAt(right) - 'A']++;
                right++;
            } else {
                charMap[s.charAt(left) - 'A']--;
                left++;
            }
        }

        sum = getSum(charMap);
        longest = Math.max(longest, sum);

        return longest;
    }

    public int getMax(int[] charMap) {
        int max = 0;
        for (int i = 0; i < 26; i++) max = Math.max(max, charMap[i]);
        return max;
    }

    public int getSum(int[] charMap) {
        int sum = 0;
        for (int i = 0; i < 26; i++) sum += charMap[i];
        return sum;
    }
}
